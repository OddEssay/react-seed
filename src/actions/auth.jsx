export const register = (email, password) => {
  return {
    type: 'AUTH_REGISTER_REQUESTED',
    email,
    password,
  }
}

export const login = (email, password) => {
  return {
    type: 'AUTH_LOGIN_REQUESTED',
    email,
    password,
  }
}

export const logout = () => {
  return {
    type: 'AUTH_LOGOUT_REQUESTED',
  }
}

export const restoreUser = ( firebaseUser ) => {
  return {
    type: 'AUTH_LOGIN_SUCCEEDED',
    result: firebaseUser,
  }
}
