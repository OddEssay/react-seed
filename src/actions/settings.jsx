export function readValueSnapshot(ref, snap) {
  console.log('readValueSnapshot',snap)
  window.snap = snap
  return {
    type: 'FIREBASE_READ_VALUE',
    data: snap.val(),
    ref
  }
}
