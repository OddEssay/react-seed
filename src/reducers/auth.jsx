const defaultState = {
  uid: null,
  displayName: null,
  photoURL: null,
  email: null,
  registrationErrorMessages: [],
  loginErrorMessages: [],
}

function auth(state = defaultState, action){
  switch(action.type) {
    case 'AUTH_REGISTER_SUCCEEDED': {
      const { uid, displayName, photoURL, email } = action.result
      const newState = {
        ...state,
        uid,
        displayName,
        photoURL,
        email,
        registrationErrorMessages: [],
      }
      return newState
    }
    case 'AUTH_REGISTER_FAILED': {
      const { email, message } = action
      const newState = {
        ...state,
        email,
        registrationErrorMessages: [ message ],
      }
      return newState
    }
    case 'AUTH_LOGIN_SUCCEEDED': {
      const { uid, displayName, photoURL, email } = action.result
      const newState = {
        ...state,
        uid,
        displayName,
        photoURL,
        email,
        loginErrorMessages: [],
      }
      return newState
    }
    case 'AUTH_LOGIN_FAILED': {
      const { email, message } = action
      const newState = {
        ...state,
        email,
        loginErrorMessages: [ message ],
      }
      return newState
    }
    default: {
      return state
    }
  }
}
export default auth
