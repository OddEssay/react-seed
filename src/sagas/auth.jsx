import 'babel-polyfill'
import firebase from 'firebase'
import { call, put } from 'redux-saga/effects'
import { takeEvery } from 'redux-saga'
import { push } from 'react-router-redux'

function* registerAuth(action){
  try {
    const email = action.email
    const password = action.password
    const doRegister = () => firebase.auth().createUserWithEmailAndPassword(email, password)

    const register = yield call(doRegister)
    yield put({ type: 'AUTH_REGISTER_SUCCEEDED', result: register })
    yield put( push('/success') )
  } catch(e) {
    yield put({ type: 'AUTH_REGISTER_FAILED', message: e.message })
  }
}

function* AuthSaga(){
  yield* takeEvery( 'AUTH_REGISTER_REQUESTED', registerAuth)
}

export default AuthSaga
