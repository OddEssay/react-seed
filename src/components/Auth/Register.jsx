import React from 'react'
import autoBind from 'react-autobind'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import * as authActions from '../../actions/auth'
import { map } from '../../helpers'

const mapStateToProps = ( state ) => {
  return {
    auth: state.auth,
  }
}
const mapDispatchToProps = (dispatch) => {
  return {
    authActions: bindActionCreators(authActions, dispatch),
  }
}
export class Register extends React.Component {
  static displayName = 'Register'

  static propTypes = {
    authActions: React.PropTypes.object.isRequired,
    auth: React.PropTypes.object.isRequired,
  }

  constructor(props) {
    super(props)
    autoBind(this)

    this.state = {
      email: '',
      password: '',
    }
  }

  handleChange(field, event) {
    const value = event.target.value
    this.setState( { [ field ]: value } )
  }

  handleRegister() {
    const { email, password } = this.state
    this.props.authActions.register( email, password )
  }

  render(){
    return (
      <div>
        Register
        <ul>
          {
            map(
              ( error, index ) => <li key={ index }>{ error }</li>
            )(this.props.auth.registrationErrorMessages)
          }
        </ul>
        <form>
          <label>Email</label>
          <input
            type='text'
            defaultValue={this.props.auth.email}
            onChange={ event => this.handleChange('email', event) }
          />

          <br />

          <label>Password</label>
          <input
            type='password'
            onChange={ event => this.handleChange('password', event) }
          />

          <br />

          <input
            type='button'
            onClick={this.handleRegister}
            value='Register'
          />

        </form>
      </div>
    )
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Register)
