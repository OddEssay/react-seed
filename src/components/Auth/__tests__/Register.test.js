import React from 'react'
import renderer from 'react-test-renderer'
import { Register } from '../Register'
import * as authActions from '../../../actions/auth'

test('Register Component without errors', () => {
  const tree = renderer.create(
    <Register authActions={ authActions }  auth={ {} }/>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Register Component with errors', () => {
  const mockErrors = [ 'The email address is already in use by another account.' ]
  const tree = renderer.create(
    <Register authActions={ authActions } auth={ { registrationErrorMessages: mockErrors } }/>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
