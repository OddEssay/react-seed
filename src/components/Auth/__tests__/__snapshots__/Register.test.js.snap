exports[`test Register Component with errors 1`] = `
<div>
  Register
  <ul>
    <li>
      The email address is already in use by another account.
    </li>
  </ul>
  <form>
    <label>
      Email
    </label>
    <input
      defaultValue={undefined}
      onChange={[Function onChange]}
      type="text" />
    <br />
    <label>
      Password
    </label>
    <input
      onChange={[Function onChange]}
      type="password" />
    <br />
    <input
      onClick={[Function bound handleRegister]}
      type="button"
      value="Register" />
  </form>
</div>
`;

exports[`test Register Component without errors 1`] = `
<div>
  Register
  <ul />
  <form>
    <label>
      Email
    </label>
    <input
      defaultValue={undefined}
      onChange={[Function onChange]}
      type="text" />
    <br />
    <label>
      Password
    </label>
    <input
      onChange={[Function onChange]}
      type="password" />
    <br />
    <input
      onClick={[Function bound handleRegister]}
      type="button"
      value="Register" />
  </form>
</div>
`;
